In [2]:
import json

import pandas as pd


In [3]:
key_order = [
    "datetime",
    "now",
    "10min",
    "1hr",
    "3hr",
    "6hr",
    "12hr",
    "24hr",
    "2day",
    "3day",
]
# format 1
# key_mapping1 = {
#    "NOW": "now",
#    "MIN_10": "10min",
#    "RAIN": "1hr",  # I assume
#    "HOUR_3": "3hr",
#    "HOUR_6": "6hr",
#    "HOUR_12": "12hr",
#    "HOUR_24": "24hr",
#    "latest_2days": "2day",
#    "latest_3days": "3day",
# }
# format 2
key_mapping = {
    "Now": "now",
    "Past10Min": "10min",
    "Past1hr": "1hr",
    "Past3hr": "3hr",
    "Past6Hr": "6hr",
    "Past12hr": "12hr",
    "Past24hr": "24hr",
    "Past2days": "2day",
    "Past3days": "3day",
}

with open("./test.json", "r") as f:


In [37]:
with open("./weather10-03_11-02.json", "r") as f:
    file_data1 = f.read()  # .lstrip().rstrip()
    data1 = json.loads(file_data1)
with open("./1029-1128.json", "r") as f:
    file_data2 = f.read()  # .lstrip().rstrip()
    data2 = json.loads(file_data2)
with open("./1109_1209.json", "r") as f:
    file_data3 = f.read()  # .lstrip().rstrip()
    data3 = json.loads(file_data3)

In [38]:
taipei_data1 = data1["cwaopendata"]["resources"]["resource"]["data"]["surfaceObs"]["location"][3]["stationObsTimes"]["stationObsTime"]
taipei_data2 = data2["cwaopendata"]["resources"]["resource"]["data"]["surfaceObs"]["location"][3]["stationObsTimes"]["stationObsTime"]
taipei_data3 = data3["cwaopendata"]["resources"]["resource"]["data"]["surfaceObs"]["location"][3]["stationObsTimes"]["stationObsTime"]

In [24]:
print(type(taipei_data1))

<class 'list'>


In [39]:


# have some duplicate (handle by pandas)
taipei_data = taipei_data1+taipei_data2+taipei_data3
print(taipei_data3[-1])

{'DateTime': '2023-12-09T24:00:00+08:00', 'weatherElements': {'AirPressure': '1010.9', 'AirTemperature': '20.6', 'RelativeHumidity': '92', 'WindSpeed': '0.8', 'WindDirection': '東南,SE', 'Precipitation': '0.0', 'SunshineDuration': None}}


In [40]:

rain_ar=[]
all_ar=[]
rain_dic={}
Tcnt = 0
Xcnt = 0
Rcnt = 0
T = 0.3 
X = None
for d in taipei_data:
    all = d["weatherElements"]
    rain = d["weatherElements"]["Precipitation"]
    datetime= d["DateTime"]
    if rain == "X":
        Xcnt += 1
        rain = X
    elif rain == "T":
        Tcnt += 1
        rain = T
    else:
        rain = float(rain)
        if rain < 0.0:
            raise ValueError("rain is negative")
        if rain > 0.0:
            Rcnt+=1
        if rain in rain_dic:
            rain_dic[rain]+=1
        else:
            rain_dic[rain]=1
    rain_ar.append({"datetime":datetime,"rain":rain})
    all= {"datetime": datetime,**all}
    all_ar.append(all)
print(f"Total: {len(taipei_data)} (dup)")
print(Tcnt)
print(Rcnt)
print(Xcnt)
print(rain_dic)

Total: 2232 (dup)
266
129
0
{0.0: 1837, 1.0: 15, 0.5: 74, 3.0: 8, 2.5: 5, 8.5: 1, 1.5: 7, 5.0: 1, 7.5: 2, 6.5: 1, 2.0: 7, 5.5: 2, 4.0: 3, 9.0: 1, 7.0: 1, 3.5: 1}


In [41]:
for e in rain_ar:
    print(e)

{'datetime': '2023-10-03T01:00:00+08:00', 'rain': 0.0}
{'datetime': '2023-10-03T02:00:00+08:00', 'rain': 0.0}
{'datetime': '2023-10-03T03:00:00+08:00', 'rain': 0.0}
{'datetime': '2023-10-03T04:00:00+08:00', 'rain': 0.0}
{'datetime': '2023-10-03T05:00:00+08:00', 'rain': 0.0}
{'datetime': '2023-10-03T06:00:00+08:00', 'rain': 0.0}
{'datetime': '2023-10-03T07:00:00+08:00', 'rain': 0.0}
{'datetime': '2023-10-03T08:00:00+08:00', 'rain': 0.0}
{'datetime': '2023-10-03T09:00:00+08:00', 'rain': 0.0}
{'datetime': '2023-10-03T10:00:00+08:00', 'rain': 0.3}
{'datetime': '2023-10-03T11:00:00+08:00', 'rain': 0.3}
{'datetime': '2023-10-03T12:00:00+08:00', 'rain': 0.3}
{'datetime': '2023-10-03T13:00:00+08:00', 'rain': 0.0}
{'datetime': '2023-10-03T14:00:00+08:00', 'rain': 0.3}
{'datetime': '2023-10-03T15:00:00+08:00', 'rain': 0.3}
{'datetime': '2023-10-03T16:00:00+08:00', 'rain': 1.0}
{'datetime': '2023-10-03T17:00:00+08:00', 'rain': 0.3}
{'datetime': '2023-10-03T18:00:00+08:00', 'rain': 1.0}
{'datetime

In [42]:
for e in all_ar:
    print(e)

{'datetime': '2023-10-03T01:00:00+08:00', 'AirPressure': '1008.2', 'AirTemperature': '24.9', 'RelativeHumidity': '80', 'WindSpeed': '1.2', 'WindDirection': '東南,SE', 'Precipitation': '0.0', 'SunshineDuration': None}
{'datetime': '2023-10-03T02:00:00+08:00', 'AirPressure': '1007.5', 'AirTemperature': '24.5', 'RelativeHumidity': '82', 'WindSpeed': '1.5', 'WindDirection': '東,SE', 'Precipitation': '0.0', 'SunshineDuration': None}
{'datetime': '2023-10-03T03:00:00+08:00', 'AirPressure': '1006.9', 'AirTemperature': '24.2', 'RelativeHumidity': '83', 'WindSpeed': '1.4', 'WindDirection': '南南東,SSE', 'Precipitation': '0.0', 'SunshineDuration': None}
{'datetime': '2023-10-03T04:00:00+08:00', 'AirPressure': '1006.8', 'AirTemperature': '24.1', 'RelativeHumidity': '82', 'WindSpeed': '1.2', 'WindDirection': '東,SE', 'Precipitation': '0.0', 'SunshineDuration': None}
{'datetime': '2023-10-03T05:00:00+08:00', 'AirPressure': '1006.9', 'AirTemperature': '24.1', 'RelativeHumidity': '77', 'WindSpeed': '1.8', '

To dataframe 

In [43]:
import pandas as pd
rain_df = pd.DataFrame(rain_ar)
rain_df.drop_duplicates(subset="datetime",inplace=True)
print(rain_df)


                       datetime  rain
0     2023-10-03T01:00:00+08:00   0.0
1     2023-10-03T02:00:00+08:00   0.0
2     2023-10-03T03:00:00+08:00   0.0
3     2023-10-03T04:00:00+08:00   0.0
4     2023-10-03T05:00:00+08:00   0.0
...                         ...   ...
2227  2023-12-09T20:00:00+08:00   0.0
2228  2023-12-09T21:00:00+08:00   0.0
2229  2023-12-09T22:00:00+08:00   0.0
2230  2023-12-09T23:00:00+08:00   0.0
2231  2023-12-09T24:00:00+08:00   0.0

[1632 rows x 2 columns]


                       datetime  rain
0     2023-10-03T01:00:00+08:00   0.0
1     2023-10-03T02:00:00+08:00   0.0
2     2023-10-03T03:00:00+08:00   0.0
3     2023-10-03T04:00:00+08:00   0.0
4     2023-10-03T05:00:00+08:00   0.0
...                         ...   ...
2227  2023-12-09T20:00:00+08:00   0.0
2228  2023-12-09T21:00:00+08:00   0.0
2229  2023-12-09T22:00:00+08:00   0.0
2230  2023-12-09T23:00:00+08:00   0.0
2231  2023-12-09T24:00:00+08:00   0.0

[1632 rows x 2 columns]

>>> 1632/(31+30+7)
24.0


In [46]:
all_df = pd.DataFrame(all_ar)
all_df.drop_duplicates(subset="datetime",inplace=True)
print(all_df)

                       datetime AirPressure AirTemperature RelativeHumidity  \
0     2023-10-03T01:00:00+08:00      1008.2           24.9               80   
1     2023-10-03T02:00:00+08:00      1007.5           24.5               82   
2     2023-10-03T03:00:00+08:00      1006.9           24.2               83   
3     2023-10-03T04:00:00+08:00      1006.8           24.1               82   
4     2023-10-03T05:00:00+08:00      1006.9           24.1               77   
...                         ...         ...            ...              ...   
2227  2023-12-09T20:00:00+08:00      1012.1           22.1               85   
2228  2023-12-09T21:00:00+08:00      1011.9           21.7               88   
2229  2023-12-09T22:00:00+08:00      1011.6           21.5               89   
2230  2023-12-09T23:00:00+08:00      1011.2           21.0               91   
2231  2023-12-09T24:00:00+08:00      1010.9           20.6               92   

     WindSpeed WindDirection Precipitation Sunshine

                       datetime AirPressure AirTemperature RelativeHumidity  \
0     2023-10-03T01:00:00+08:00      1008.2           24.9               80   
1     2023-10-03T02:00:00+08:00      1007.5           24.5               82   
2     2023-10-03T03:00:00+08:00      1006.9           24.2               83   
3     2023-10-03T04:00:00+08:00      1006.8           24.1               82   
4     2023-10-03T05:00:00+08:00      1006.9           24.1               77   
...                         ...         ...            ...              ...   
2227  2023-12-09T20:00:00+08:00      1012.1           22.1               85   
2228  2023-12-09T21:00:00+08:00      1011.9           21.7               88   
2229  2023-12-09T22:00:00+08:00      1011.6           21.5               89   
2230  2023-12-09T23:00:00+08:00      1011.2           21.0               91   
2231  2023-12-09T24:00:00+08:00      1010.9           20.6               92   

     WindSpeed WindDirection Precipitation SunshineDuration  
0          1.2         東南,SE           0.0             None  
1          1.5          東,SE           0.0             None  
2          1.4       南南東,SSE           0.0             None  
3          1.2          東,SE           0.0             None  
4          1.8          東,SE           0.0             None  
...        ...           ...           ...              ...  
2227       2.2           東,E           0.0             None  
2228       1.8           東,E           0.0             None  
2229       1.4       東南東,ESE           0.0             None  
2230       0.9          東,SE           0.0             None  
2231       0.8         東南,SE           0.0             None  

[1632 rows x 8 columns]

In [47]:


rain_df.to_pickle("./10-03_12_09_rain.pkl")
all_df.to_pickle("./10-03_12_09_all.pkl")